text/0000-yarn-berry-support.md

```diff
@@ -0,0 +1,65 @@
+
+# Pedagogy
+[pedagogy]: #pedagogy
+
+The ideal outcome with this change is that there will be very little we need to teach users. Newer versions of Yarn will be supported and users will be able to continue to use Volta the way they always have, without needing to worry about specific details of their setup.
+
+The main detail we will need to communicate is the lack of support for the 2.x major version of Yarn (only supporting Yarn 1.x and Yarn 3.0+).
+
+# Details
+[details]: #details
+
+## Skip Yarn 2.x Support
+
+As mentioned above, the initial release of Yarn Berry involved a lot of changes that were difficult to track and didn't fit with Volta's model for handling package installation. The biggest blocker is that the distribution package on npm wasn't created until the last version of Yarn 2. Before that, there wasn't any reliable way to fetch the Yarn binary and make it available within Volta.
+
+Since the only viable distribution is the _final_ version of Yarn 2 and Yarn has since released Yarn 3, it makes the most sense to skip support for Yarn 2 alltogether and jump straight from Yarn 1 to Yarn 3. That will allow us to detect the supported versions very easily, using only the major version number rather than a bespoke list of "allowed" versions. We should provide a useful error message if a user attempts to install or pin Yarn 2, directing them instead to use Yarn 3 for the most up-to-date experience.
+
+## Installation
+
+Currently, for Yarn 1, we fetch the tarball from the [`yarn`](https://www.npmjs.com/package/yarn) package on the npm registry. However, Yarn Berry isn't included in that package, so we instead need to fetch the package at [`@yarnpkg/cli-dist`](https://www.npmjs.com/package/@yarnpkg/cli-dist). That package includes everything we need for Volta's drop-in install: The JS script and the executables to launch them. When fetching a version of Yarn, we will need to detect the version and choose the appropriate package, then download and install it using the existing logic.
+
+### Hooks
+
+One outstanding question is how we want to handle the hooks for installs. Since the Hooks format was created before we switched Yarn to use the npm registry, the logic is rather complicated:
+
+* If we aren't using hooks, then we fetch from npm and expect the data to match the npm package format
+* If we _are_ using hooks, then we fetch from the location specified by the hook and we expect the data to be in the old GitHub API format
+
+With the addition of Yarn 3 pulling from a different package (though still using the npm registry), this could get more complicated. My current suggestion is that we keep the hooks format the same even for Yarn Berry releases. This will mean that teams using hooks will have to make sure the correct package is available at the appropriate URLs themselves. Then, in a future change where we update the hooks format, we can migrate all of Yarn to use the npm format and normalize the behavior.
```

Comment:
A small change to help with this could be an optional `"format"` field in the `"yarn"` section of hooks.json, which would take `"npm" | "github"`, like:

```
    "yarn": {
        "index": {
            "template": "http://example.com/yarn/{{os}}/{{arch}}/yarn-{{version}}.tgz"
        },
        "latest": {
            "prefix": "http://example.com/yarnpkg/"
        },
        "distro": {
            "bin": "~/yarn-distro"
        },
        "format": "npm"
    }
```

That would specify the format Volta should expect. Leaving this out would default to the current behavior of using the Github API format.

Reply:
Great suggestion! That would make the different formats more clear / explicit in a backwards-compatible way. It would also enable teams to use the `npm` format for all versions of Yarn, even the older 1.x ones.